Add REX Pull Client to Project Server (#2206)
theforeman/foreman-documentation: guides/common/modules/proc_configuring-remote-execution-for-pull-client-project-server.adoc

[id="configuring-remote-execution-for-pull-client-project-server_{context}"]

> = Configuring Remote Execution for Pull Client on the {Project} Server

Reviewer: ```suggestion
= Configuring Remote Execution for Pull Client on {ProjectServer}
```
see https://github.com/theforeman/foreman-documentation/blob/master/guides/common/attributes-base.adoc#L99

However, Remote Execution also offers pull-based transport, which you can use if your infrastructure prohibits outgoing connections from {Project} to hosts.

This comprises `pull-mqtt` mode on {Project} in combination with a pull client running on hosts.
Configure the `pull-mqtt` mode to migrate from Katello Agent, which is a deprecated method of pull-based transport.

[NOTE]
====
The `pull-mqtt` mode works only with the Script provider.
Ansible and other providers will continue to use their default transport settings.
====

To use `pull-mqtt` mode on the {Project}, follow the procedure below:

.Procedure
. Enable the pull-based transport on each relevant {ProjectServer}:
+
[options="nowrap" subs="quotes,attributes"]
----
 # {installer-scenario} \
--foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
----
. Configure the firewall to allow MQTT service on port 1883:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# firewall-cmd --add-port="1883/tcp"
# firewall-cmd --runtime-to-permanent
----
. In `pull-mqtt` mode, hosts subscribe for job notifications to the {Project} (or {SmartProxy}) through which they are registered.

> Therefore, it is recommended to ensure that {ProjectServer} sends remote execution jobs to that same {Project} (or {SmartProxy}).

Reviewer 2: I feel like with numbered lists in procedures, we should try and make each step one sentence instead of multiple sentences unless of course you reference another document or section. Should this be broken down? I'm wondering if `In pull-mqtt mode, hosts subscribe for job notifications to the {Project} (or {SmartProxy}) through which they are registered. Therefore, it is recommended to ensure that {ProjectServer} sends remote execution jobs to that same {Project} (or {SmartProxy}).` should be their own paragraphs and then the numbered list continues with `. In the {ProjectWebUI}, navigate to Administer > Settings. 
. On the Content tab, set the value of Prefer registered through {SmartProxy} for remote execution to Yes.` followed by 
`. After you set up the pull-based transport on {Project}, you must also configure it on each host. 
. For more information, see {ManagingHostsDocURL}transport-modes-for-remote-execution_managing-hosts[Transport Modes for Remote Execution] in {ManagingHostsDocTitle}.`

Any thoughts?

Reviewer 2: It would end up looking like this but with the attributes rendered, of course:
![image](https://github.com/theforeman/foreman-documentation/assets/91690569/1c424fb5-5424-4195-b0f5-049c3727c1f1)

Author: Agree to this :+1: 
Even applicable to proxy section but will cover in other PR.

> [id="configuring-remote-execution-for-pull-client-project-server_{context}"]
> = Configuring Remote Execution for Pull Client on the {ProjectServer}

Reviewer: ```suggestion
[id="configuring-remote-execution-for-pull-client-on-{project-context}-server_{context}"]
= Configuring Remote Execution for Pull Client on {ProjectServer}
```

Please also add "on" to the file name.

By default, Remote Execution uses SSH as the transport mechanism for the Script provider.
However, Remote Execution also offers pull-based transport, which you can use if your infrastructure prohibits outgoing connections from {Project} to hosts.

This comprises `pull-mqtt` mode on {Project} in combination with a pull client running on hosts.

> Configure the `pull-mqtt` mode to migrate from Katello Agent, which is a deprecated method of pull-based transport.

Reviewer: Maybe add "If you still use Katello Agent, configure ...".

Author: Reworded :+1:

[NOTE]
====
The `pull-mqtt` mode works only with the Script provider.
Ansible and other providers will continue to use their default transport settings.
====

To use `pull-mqtt` mode on the {Project}, follow the procedure below:

.Procedure

> . Enable the pull-based transport on each relevant {ProjectServer}:

Reviewer: ```suggestion
. Enable the pull-based transport on your {ProjectServer}:
```

+
[options="nowrap" subs="quotes,attributes"]
----
 # {installer-scenario} \
--foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
----
. Configure the firewall to allow MQTT service on port 1883:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# firewall-cmd --add-port="1883/tcp"
# firewall-cmd --runtime-to-permanent
----
+

> In `pull-mqtt` mode, hosts subscribe for job notifications to the {Project} (or {SmartProxy}) through which they are registered.

Reviewer: ```suggestion
In `pull-mqtt` mode, hosts subscribe for job notifications to either your {Project} or any {SmartProxyServer} through which they are registered.
```

Or just say "... to {SmartProxy} through ..."
